// easy-connect/dist/util/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var options_1 = require("./options");
exports.construct = options_1.construct;
exports.constructArray = options_1.constructArray;
exports.createApiData = options_1.createApiData;
exports.updateApiData = options_1.updateApiData;
exports.deleteApiData = options_1.deleteApiData;
exports.genInstance = options_1.genInstance;
var util_1 = require("./util");
exports.authConfig = util_1.authConfig;
exports.api = util_1.api;
exports.query = util_1.query;
exports.id = util_1.id;
exports.cacheable = util_1.cacheable;
exports.secure = util_1.secure;
exports.create = util_1.create;
exports.access = util_1.access;
exports.accessQuery = util_1.accessQuery;
exports.isSecure = util_1.isSecure;
exports.isCacheable = util_1.isCacheable;
exports.getName = util_1.getName;
exports.accessId = util_1.accessId;
exports.getBaseUrl = util_1.getBaseUrl;
exports.accessFilter = util_1.accessFilter;
exports.config = util_1.config;
exports.key = util_1.key;
exports.isPrimitive = util_1.isPrimitive;
exports.getQueryKey = util_1.getQueryKey;
