// easy-connect/dist/fetch/store.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const Observable_1 = require("rxjs/Observable");
require("rxjs/add/observable/from");
require("rxjs/add/operator/do");
const util_1 = require("../util");
const easy_injectionjs_1 = require("easy-injectionjs");
let Store = class Store {
    constructor() {
        this.contents = new class {
        }();
    }
    getData(key) {
        return Observable_1.Observable.of(this.contents[util_1.getName(key)]);
    }
    getDataByKey(key, id) {
        return Observable_1.Observable.of(this.contents[util_1.getName(key)][id]);
    }
    putData(key, data) {
        return data.do(x => {
            this.contents[util_1.getName(key)] = x;
        });
    }
    putDataByKey(key, data, id) {
        return data.do(x => {
            this.contents[util_1.getName(key)][id] = x;
        });
    }
    check(key) {
        return this.contents[util_1.getName(key)] ? true : false;
    }
    checkByKey(key, id) {
        return this.contents[util_1.getName(key)][id]
            ? true : false;
    }
};
Store = __decorate([
    easy_injectionjs_1.EasySingleton('STORE'),
    __metadata("design:paramtypes", [])
], Store);
exports.Store = Store;
