// easy-connect/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var util_1 = require("./util");
exports.authConfig = util_1.authConfig;
exports.api = util_1.api;
exports.id = util_1.id;
exports.query = util_1.query;
exports.secure = util_1.secure;
exports.cacheable = util_1.cacheable;
exports.key = util_1.key;
var core_1 = require("./core");
exports.EasyConnect = core_1.EasyConnect;
exports.container = core_1.container;
exports.All = core_1.All;
exports.Add = core_1.Add;
exports.Query = core_1.Query;
exports.Update = core_1.Update;
exports.Delete = core_1.Delete;
exports.AuthType = core_1.AuthType;
exports.Get_All = core_1.Get_All;
exports.Get_Query = core_1.Get_Query;
exports.Get_Current = core_1.Get_Current;
exports.Current = core_1.Current;
var auth_1 = require("./auth");
exports.Token = auth_1.Token;
exports.HttpFactory = auth_1.HttpFactory;
exports.EasyTokenAuth = auth_1.EasyTokenAuth;
exports.EasyAuth = auth_1.EasyAuth;
exports.AuthFactory = auth_1.AuthFactory;
exports.Login = auth_1.Login;
exports.Logout = auth_1.Logout;
exports.Register = auth_1.Register;
exports.Validate = auth_1.Validate;
exports.ValidateData = auth_1.ValidateData;
var easy_injectionjs_1 = require("easy-injectionjs");
exports.EasyPrototype = easy_injectionjs_1.EasyPrototype;
exports.EasyFactory = easy_injectionjs_1.EasyFactory;
exports.EasySingleton = easy_injectionjs_1.EasySingleton;
exports.Easy = easy_injectionjs_1.Easy;
